solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -484,6 +498,12 @@ private void refreshLegacyClusterState(Watcher watcher)
         }
         this.legacyCollectionStates = loadedData.getCollectionStates();
         this.legacyClusterStateVersion = stat.getVersion();
+        for (Map.Entry<String, ClusterState.CollectionRef> entry : this.legacyCollectionStates.entrySet()) {
+          if (entry.getValue().isLazilyLoaded() == false) {
+            // a watched collection - trigger notifications
+            notifyStateWatchers(entry.getKey(), entry.getValue().get());
+          }
+        }
```

Comment:
I think it would add a lot of value here to actually check differences.  There really wouldn't be much computational work since you could restrict it to watched collections.  Something like:

```
        for (Map.Entry<String, CollectionWatch> watchEntry : this.collectionWatches.entrySet()) {
          String coll = watchEntry.getKey();
          CollectionWatch collWatch = watchEntry.getValue();
          DocCollection newState = this.legacyCollectionStates.get(coll).get();
          if (!collWatch.stateWatchers.isEmpty()
              && !Objects.equals(oldCollectionStates.get(coll).get(), newState)) {
            notifyStateWatchers(coll, newState);
          }
        }
```